exchanges/bybit/bybit_websocket.go

```diff
@@ -0,0 +1,491 @@
+				if err != nil {
+					return err
+				}
+
+				p, err := by.extractCurrencyPair(data.Parameters.Symbol, asset.Spot)
+				if err != nil {
+					return err
+				}
+
+				side := order.Sell
+				if data.TradeData.Side {
+					side = order.Buy
+				}
+
+				return trade.AddTradesToBuffer(by.Name, trade.Data{
+					Timestamp:    data.TradeData.Time.Time(),
+					CurrencyPair: p,
+					AssetType:    asset.Spot,
+					Exchange:     by.Name,
+					Price:        data.TradeData.Price,
+					Amount:       data.TradeData.Size,
+					Side:         side,
+					TID:          data.TradeData.ID,
+				})
+			case wsTicker:
+				var data WsSpotTicker
+				err := json.Unmarshal(respRaw, &data)
+				if err != nil {
+					return err
+				}
+
+				p, err := by.extractCurrencyPair(data.Ticker.Symbol, asset.Spot)
+				if err != nil {
+					return err
+				}
+
+				by.Websocket.DataHandler <- &ticker.Price{
+					ExchangeName: by.Name,
+					Bid:          data.Ticker.Bid,
+					Ask:          data.Ticker.Ask,
+					LastUpdated:  data.Ticker.Time.Time(),
+					AssetType:    asset.Spot,
+					Pair:         p,
+				}
+				return nil
+			case wsKlines:
+				var data KlineStream
+				err := json.Unmarshal(respRaw, &data)
+				if err != nil {
+					return err
+				}
+
+				p, err := by.extractCurrencyPair(data.Kline.Symbol, asset.Spot)
+				if err != nil {
+					return err
+				}
+
+				by.Websocket.DataHandler <- stream.KlineData{
+					Pair:       p,
+					AssetType:  asset.Spot,
+					Exchange:   by.Name,
+					StartTime:  data.Kline.StartTime.Time(),
+					Interval:   data.Parameters.KlineType,
+					OpenPrice:  data.Kline.OpenPrice,
+					ClosePrice: data.Kline.ClosePrice,
+					HighPrice:  data.Kline.HighPrice,
+					LowPrice:   data.Kline.LowPrice,
+					Volume:     data.Kline.Volume,
+				}
+				return nil
+			default:
+				by.Websocket.DataHandler <- stream.UnhandledMessageWarning{Message: by.Name + stream.UnhandledMessage + string(respRaw)}
+			}
+		}
+
+		if m, ok := d["auth"]; ok {
+			log.Infof(log.WebsocketMgr, "%v received auth response: %v", by.Name, m)
+			return nil
+		}
+
+		if m, ok := d["ping"]; ok {
+			log.Infof(log.WebsocketMgr, "%v received ping: %v", by.Name, m)
+			return nil
+		}
+	case []interface{}:
+		if obj, ok := d[0].(map[string]interface{}); ok {
+			if e, ok := obj["e"].(string); ok {
+				switch e {
+				case wsAccountInfoStr:
+					var data []wsAccountInfo
+					err := json.Unmarshal(respRaw, &data)
+					if err != nil {
+						return fmt.Errorf("%v - Could not convert to outboundAccountInfo structure %s",
+							by.Name,
+							err)
+					}
+					by.Websocket.DataHandler <- data[0]
+					return nil
+				case wsOrderStr:
+					var data []wsOrderUpdate
+					err := json.Unmarshal(respRaw, &data)
+					if err != nil {
+						return fmt.Errorf("%v - Could not convert to executionReport structure %s",
+							by.Name,
+							err)
+					}
+					oType, err := order.StringToOrderType(data[0].OrderType)
```

Comment:
Let's put this in a for loop to prevent any potential data loss from multiple orders in one update, or empty slice panics